src/modules/previewpane/MonacoPreviewHandler/languages.json

```diff
@@ -431,23 +431,6 @@
         "m3"
       ]
     },
-    {
```

Comment:
Don't change this file!

This extensions won't get registred when they get registred by the markdown handler anyway. Right @stefansjfw?

Reply:
Question: You are saying that simply removing `languages.json` from my PR and doing nothing else has no impact on the result? (It would be great, though!) Or am I expected to do something else?

Reply:
I'm struggling with File Explorer crashes on my test VM while I toggle the preview handlers on and off. Would you please tell me what's the idea behind this review and keeping the `languages.json`?

And I mishandled my forked repo. Now there is noise in this PR.

Reply:
I agree with @Aaron-Junker, this file is generated, we should not change it manually.

Reply:
@stefansjfw I can eliminate the need to use `languages.json` altoghter. Is that acceptable?

Reply:
Yes, I can elaborate. `languages.json` has only one purpose in `modulesRegistry.h`: Provide us with a list of filename extensions to register under `HKEY_CURRENT_USER\Software\Classes`. The file, however, is not optimized for this simple task. It is bloated with features you don't need, like MIME types, and both of my code reviewers (you and Aaron-Junker) are afraid of touching it.

Since this file is so scary, why use it in the first place? We have two options.

**OPTION 1:** In `modulesRegistry.h`, we use a developer-curated `PreviewHandlerTypes.js`:

```json
[
  {
    "handler": "MarkdownPreviewHandler",
    "extensions": [
      "md",
      "markdown",
      ".mdown",
      ".mkdn",
      ".mkd",
      ".mdwn",
      ".mdtxt",
      ".mdtext"
    ]
  },
  {
    "handler": "SvgPreviewHandler",
    "extensions": [
      "svg"
    ]
  },
  {
    "handler": "PdfPreviewHandler",
    "extensions": [
      "pdf"
    ]
  },
  {
    "handler": "GcodePreviewHandler",
    "extensions": [
      "gcode"
    ]
  },
  {
    "handler": "MonacoPreviewHandler",
    "extensions": [
      "All those file name extensions come here, minus Markdown, SVG, SVG-Z, PDF, GCODE, Plain Text, INI, and everything Monaco doesn't handle well."
    ]
  }
]
```

**OPTION 2:** Forgo any dependency on JSON files and enter type names directly within `modulesRegistry.h`.

Reply:
@skycommand
1. This file is also used by the FileHandler class to get the matching id to the extension
2. This file is an auto generated file. If you change something in this file it gets overwirtten the next time someone regenerates it.

Reply:
Aaron: Interest in any of the alternatives I suggested above?

Reply:
@skycommand Option 1: why? Why should we create another file when all the information needed is already in a file
Option 2: Too elaborate.

As far as I understand this if you register these file types for the markdown preview handler, they won't get registred for the dev file handler.

Reply:
I don't see why should we manually write and maintain bunch of extensions for MonacoPreviewHandler if we have that automatically generated by the engine we use in the background to show the preview in the first place?

Reply:
Alright, gentlemen, let's take it from the top, making sure that we are on the same page.

### First, our problem

Currently, as of PowerToys 0.55.2, file name extension `.md` is registered with `MarkdownPreviewHandler` while `    .markdown`, `.mdown`, `.mkdn`, `.mkd`, `.mdwn`, `.mdtxt`, and `.mdtext` are registered with the inferior `MonacoPreviewHandler`.

### My proposals to solve the problem

1. Rewrite `languages.json` to include only file name extensions that we want to register with `MonacoPreviewHandler`.
2. Write a `PreviewHandlerTypes.json` that contains only the file name extensions that `MonacoPreviewHandler` must register. I labeled it "OPTION 1" above.
3. Forgo any dependency on JSON files and enter type names directly within `modulesRegistry.h`. I labeled it "OPTION 2" above.
4. Continue with the current approach, i.e. read from `languages.json` but have `modulesRegistry.h` filter it.

### Objections to my proposal

Now, you gentlemen (my esteemed colleagues, Aaron Junker and Stefan Markovic) have rejected all my four proposals:

1. Proposal 1 has been rejected because `languages.json` is automatically generated and has other uses besides registering preview handlers.

2. Proposal 2 has been rejected because:

    > Why should we create another file when all the information needed is already in a file? (Aaron-Junker)

    > I don't see why should we manually write and maintain [a] bunch of extensions for MonacoPreviewHandler if we have that automatically generated by the engine we use in the background to show the preview in the first place? (Stefan Markovic)

    To be clear: No, we **do not** have all the information needed in one file. `Languages.json` contains only 50% of the information we need: File types that _Monaco can handle_. We also need another piece of information: File types that we _want Monaco to handle_. They are in `modulesRegistry.h`, lines 17, 18, and 72–77.

3. Proposal 3 has been rejected because:

    > Too elaborate (Aaron-Junker)

    True, but you've already rejected all the less elaborate approaches. Stefan Markovic has made no comments.

4. Proposal 4 has been rejected because:

    > This will break the File explorer add-ons on/off logic. (Stefan Markovic)

    This is vague, I'm afraid. Neither I nor anybody else can proceed without additional info, such as a screenshot showing the breakage or other manifestations/evidence.

    Aaron Junker has not commented on this proposal.

### So, how do we move forward?

The above describes our impasse so far. To get past this deadlock, I'd like to ask a few questions from my esteemed colleagues.

1. Do you agree that we have a problem? (This one is more of a rhetorical question than an actual question. Both of you have already said "yes" and I thank you for that.)
2. Do you agree that to solve the problem, something must inevitably change?
3. What is your proposal to fix this problem? I cannot come up with more ideas on my own, I'm afraid.

-----

According to Grammarly, this message is Confident, Friendly, and Optimistic. The friendly and confident part was my purpose. As for the optimistic part, it appears that Grammarly has faith that we do not reach an impasse. 😊

Reply:
What I tried to make clear is that Lines 16 and 17 (https://github.com/microsoft/PowerToys/blob/main/src/common/utils/modulesRegistry.h#L16) prevented the installation of these file types for the source code previewer. The only zjing necessary would be to revert your deletion of these lines and extend them to check for the newly added markdown extensions.

Reply:
I'll be happy to answer your questions and to get to the same page:
1. Yes, the problem is there and Yes I don't like how it's handled at the moment.
2. Obviously.
3. As Monaco is superior to other PreviewHandlers (it handles many extensions related or non-related, opposite to other handler that handle just a few of related extensions) we can expect it to get in conflict with potential new handlers. That means we have to elegantly limit Monaco to stop processing extensions that current (and potential new) handlers are handling better. On the other hand, as I already said maintaining manually written JSON file this big is something I'd like to avoid tbh. As that can be generated by monaco, let's leave it as it is. Also, Monaco engine might evolve in the future and if we want to keep up with it, the way forward is as it is now.

So, to come back to the possible solutions.. 
I agree with you on "languages.json does NOT contain ALL the info we need". So, to come back to Monaco being (and probably staying) superior handler, I'd go with the solution that will only limit Monaco and not affecting other handlers. Something like - adding 'ignore-extensions.json' file, which will contain extensions that should be ignored by Monaco.. I understand that this is double-work in sense of reading extensions from lanugages.json and then ignoring some of the extensions we know are there, but I think I already said enough times why I think keeping the languages.json as it is is a good ide :) Looking forward to hear what do you think about this solution?

Reply:
Well, this round of conversation was a huge step forward:

- Aaron Junker, you have unambiguously sided with Proposal 4.
- Stefan Markovic, you are in favor of a variant of Proposal 4 that involves a second JSON file. I like this version.

I haven't fully made up my mind, but the JSON file might look like this:

```json
[
  {
    "handler": "MarkdownPreviewHandler",
    "extensions": [ "md", "markdown", "mdown", "mkdn", "mkd", "mdwn", "mdtxt", "mdtext" ]
  },
  {
    "handler": "SvgPreviewHandler",
    "extensions": [ "svg" ]
  },
  {
    "handler": "PdfPreviewHandler",
    "extensions": [ "pdf" ]
  },
  {
    "handler": "GcodePreviewHandler",
    "extensions": [ "gcode" ]
  }
]
```

### Immediate plan of action

Here is what I am planning to do:

1. I'll mark these reviews as "Resolved" and revert changes to `modulesRegistry.h` and `languages.json` in my repo.
2. I'll start working on a revised `modulesRegistry.h` that implements Stefan Markovic's variant of Proposal 4.

Do I need to do something like moving this PR to – I don't know – draft status or something? Or can it be open for a few more days? It's my understanding that a bot would mark it as stale or something.

Reply:
> Well, this round of conversation was a huge step forward:
> 
> - Aaron Junker, you have unambiguously sided with Proposal 4.
> - Stefan Markovic, you are in favor of a variant of Proposal 4 that involves a second JSON file. I like this version.
> 
> I haven't fully made up my mind, but the JSON file might look like this:
> 
> ```json
> [
>   {
>     "handler": "MarkdownPreviewHandler",
>     "extensions": [ "md", "markdown", "mdown", "mkdn", "mkd", "mdwn", "mdtxt", "mdtext" ]
>   },
>   {
>     "handler": "SvgPreviewHandler",
>     "extensions": [ "svg" ]
>   },
>   {
>     "handler": "PdfPreviewHandler",
>     "extensions": [ "pdf" ]
>   },
>   {
>     "handler": "GcodePreviewHandler",
>     "extensions": [ "gcode" ]
>   }
> ]
> ```
> 
> ### Immediate plan of action
> 
> Here is what I am planning to do:
> 
> 1. I'll mark these reviews as "Resolved" and revert changes to `modulesRegistry.h` and `languages.json` in my repo.
> 2. I'll start working on a revised `modulesRegistry.h` that implements Stefan Markovic's variant of Proposal 4.
> 
> Do I need to do something like moving this PR to – I don't know – draft status or something? Or can it be open for a few more days? It's my understanding that a bot would mark it as stale or something.

Sounds good to me. You don't have to change it to draft status, but you can if you want. The bot doesn't mark PRs as stale.